### DISCUSSION TITLE
(1.x series) Public header files reference BIGNUM (AKA bignum_st) but do not say how big it is.  Anything trying to use OpenSSL cannot build against this.

### DISCUSSION THREAD

**Discussion by gsteemso** [openssl/openssl#21975] [Q&A] [answered]
What it says.  Anything I try to build (currently, against OpenSSL 1.1.1, but I see reports on the net that it also happens with earlier releases) chokes on an "incomplete type" because the compiler cannot determine how much memory a BIGNUM takes up.  OpenSSL itself compiles just fine, because the internal headers fully define the structure – a detail which caused me considerable bewilderment until I unpacked the tarball again to see what it looked like in its pre-assembled state.

I dislike, on principle, the need for an aftermarket patch which enables other software to compile by re-inserting this private structure definition into the public header.  However, the first member of the structure changes size according to the build environment, so I can't think of any other way to do it.

I'm not sure how you would go about fixing this _properly_, but it's pretty darned important!

**Comment by paulidale** [comment, marked as answer]
Moving many structures from the public header into private ones was **deliberate** and was the major change between 1.0.2 and 1.1.0.  We've got a [Wiki page describing the reasons](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes).

So, no this is not going to be _fixed_ because it isn't view as a problem.  All of our public APIs pass BIGNUMs by reference, so publicly knowing the size isn't required.

The fix is to update _your_ code to use `BN_new()` and `BN_free()` to dynamically allocate any BIGNUMs.

**Comment by gsteemso** [comment]
I fully agree with that and the reasoning behind it.  However, I have no control over the assorted software that tries to build against your own.  The compiler-error statements I'm getting are kind of dense and difficult to decipher, but it _looks_ like the complaint is about "how big is that thing" rather than about "where are the fields inside that thing".  I could be wrong, of course.

Weirdly, I'm getting the same sort of errors regarding `struct bignum_ctx` – but that one doesn't even appear in the internal headers!  I have no idea what to make of that.  How the heck did OpenSSL even successfully compile with that going on?

**Reply by paulidale** [reply]
`struct bignum_ctx` is defined in [crypto/bn/bn_ctx.c](https://github.com/openssl/openssl/blob/c1673a60e40f6dcd110d1a4ff3e11a3297ada2da/crypto/bn/bn_ctx.c#L63-L78).

It didn't need to be exposed to the rest of OpenSSL, so it wasn't.  Just because something was public, doesn't mean it's in the internal headers anymore.

**Comment by gsteemso** [comment]
Oh, that makes sense.  Thank you for explaining it.  (You may be able to tell that I don't do a lot with C.)

In any case, clearly there's nothing you or I can do about other people's software, so I'm marking this "answered".
